### ISSUE TITLE
Feature: Laravel 5.2 Custom Authentication Guard and Driver

### ISSUE THREAD

**Issue by mtpultz** [tymondesigns/jwt-auth#513]
The docs indicate it is possible to create your own implementation of Illuminate\Contracts\Auth\Guard and registering it as a driver in a service provider.  

I was reading about the new stateless token authentication that was added in 5.2 in a [JacobBennett Gist](https://gistlog.co/JacobBennett/090369fbab0b31130b51) (the docs are really vague), but it doesn't appear to be the same as JWT tokens.  That said it would be amazing to be able to leverage Laravel's API the same way.

Would it be possible to create a custom driver to reduce the amount of changes required to implement JWT tokens, and reduce the API a bit so using more of Laravel's API?  For example getting a user is `Auth::guard('api')->user();` using the API guard, and the equivalent could be `Auth::guard('jwt')->user();`

**Comment by daviestar** [comment]
+ 1

**Comment by tdhsmith** [comment]
This is already being worked on on the develop branch. (cf. [JWTGuard.php](https://github.com/tymondesigns/jwt-auth/blob/develop/src/JWTGuard.php))

If you're interested, there are some discussions on the matter at #376, #384, and #479, among others.

**Comment by mtpultz** [comment]
Thanks @tdhsmith, I'm very interested.  Thanks for the links.  Are there any areas that you need help on to make this usable for early stages of development using Laravel 5.2?  From the links it appears that Laravel 5.2 has a lot of methods implemented, but I'm not familiar enough with it to tell how far along it is.  Most of the issues are related to Lumen integration.  Trying to time this with our needs of upcoming development.

**Comment by mtpultz** [comment]
@tdhsmith are there steps to setup the use of the JWTGuard to test it out?

**Comment by tdhsmith** [comment]
> Are there any areas that you need help on to make this usable for early stages of development using Laravel 5.2?

Well I haven't personally worked on any of the new guard stuff, so I don't have a good sense myself. Perhaps @tymondesigns has suggestions. I think the main goals are just to test it thoroughly (are there unit tests for everything?) and then evaluate which other methods on `SessionGuard` might be useful to adapt for it.

> are there steps to setup the use of the JWTGuard to test it out?

Not yet. I think it should be sufficient to set your auth driver to `jwt` (assuming you've already done the config stuff in the docs). Then in general you should replace calls to `JWTAuth` with calls to `Auth` itself, and explore the new helpers this provides. In particular you should test out "traditional" auth sugars like `Auth::user()`, `Auth::guest()`, `Auth::id()`, and `Auth::logout()`. (If you're working in an app that already accomplishes everything with middleware route gating though, you probably won't see big changes then, since you probably aren't running any checks like this anywhere.)

**Comment by tdhsmith** [comment]
These are some other thoughts I've had on ways forward. They might not be exactly related to your question, but I figured I would share them here as long as I have them:

- There're some interesting questions on how the auth provider should be adapted, because right now there is an odd cycle in the call stack `JWTGuard --> JWT --> auth provider --> JWTGuard --> user provider`. In any case, I think it's probably dangerous to keep using the same provider class, `Illuminate`, for both guards. The guards inherently behave different and we shouldn't limit ourselves by expecting them to both conform to one provider. (I've argued previously that I don't think we should keep the `once` method names in `JWTGuard` because they imply an alternative that doesn't exist in a stateless auth system, but that might only be opinionated bikeshedding on my behalf.)

- In an ideal world, there would be an sibling repo to this that has example applications for different Laravel and Lumen versions and can also double as a base for functional tests. A number of recent issues, including the Lumen problems that eventually sparked the guard, have been caused by situations that couldn't really be detected by pure unit tests, so having whole app architectures to run would really help with some of the subtler specification/functionality issues.

**Comment by tymondesigns** [comment]
@tdhsmith on your first point, I removed the `JWTAuth` dependency from the `JWTGuard` so there is no weird auth call stack now, only `JWTGuard --> JWT --> Laravel's AuthManger`. Or did I miss something?

And you second point, It has been on my mind for a while, that it would be great to have concrete examples where people can see how things plumb together. Once I get this next release out of the way, that will be next I think.

**Comment by tdhsmith** [comment]
Nope you're totally right. I was simultaneously looking at pre-guard code and the new code, and not keeping them separate in my mind. The JWT / JWTAuth split keeps that cycle stuff from happening. Smart choice! :+1:

**Comment by daviestar** [comment]
This is the L5.1 demo app that led me here: https://laracasts.com/discuss/channels/laravel/starter-application-vuejs-laravel-dingo-jwt

**Comment by mtpultz** [comment]
I haven't been able to make this work yet using a set up that if the middleware group web is used appears to work.

**Setup Steps**
1. Removed the old provider in app.php that I had added earlier today and replaced it with the new `Tymon\JWTAuth\Providers\LaravelServiceProvider::class,`
1. Removed the old facades as they aren't required if your using JWTGuard
1. Artisan'ed in the updated config file
1. Replaced `token` with `jwt` for the API guard in auth.php, but left web as the default guard as this application will eventually use both
1. Added an `api` prefixed route group with the guard middleware `auth:api` with some free routes via `Route::auth;`, and set redirectTo `/api/v1/restricted/test`

But, when I tried to hit the URL using Postman it doesn't seem to enter into the login route, which makes me think it isn't related to JWT-Auth except if I change the middleware to `web` from `auth:api` it works.

```
Route::group([ 
    'prefix' => 'api/v1',
    'middleware' => 'auth:api', // indicate routes use API guard
], function () {

    Route::auth();

    Route::group([
        'prefix' => 'restricted',
        //'middleware' => 'auth | jwt.auth | check manually in controllers using JWT API'
    ], function () {

        Route::get('/test', function() { return 'authenticated'; });
    });
});
```

Stubbing through a couple classes to see what is output auth:web logs as I would expect

```
[2016-02-17 02:18:51] local.INFO: AuthManager::guard() $name=web  
[2016-02-17 02:18:52] local.INFO: AuthenticatesUsers::login()  
[2016-02-17 02:18:52] local.INFO: AuthenticatesUsers::login() $credentials= {"username":"mtpultz","password":"secret"} 
[2016-02-17 02:18:52] local.INFO: AuthManager::guard() $name=web  
```

But, auth:api only logs from AuthManager then throws `Sorry, the page you are looking for could not be found.`

```
[2016-02-17 02:18:51] local.INFO: AuthManager::guard() $name=api
```

Wasn't sure if `auth` or `jwt.auth` middleware was required on restricted routes was going to try and figure that out once I got a user token using it in Postman to hit the restricted route.

I'm sure it is something very stupid that I'm doing, but I haven't been able to figure out what that is yet.

**Comment by mtpultz** [comment]
This is just a post of the steps to get the [JWTGuard](https://github.com/tymondesigns/jwt-auth/blob/develop/src/JWTGuard.php) in place for an API using Laravel 5.2.x to possibly help with starting the documentation of this feature, but also to see if this is the best way to implement JWTGuard.

Using JWTGuard with Laravel 5.2.x
=====

1. `config/app.php` - add `Tymon\JWTAuth\Providers\LaravelServiceProvider::class` to providers
2. `config/auth.php` - set the default guard to `api`, and change the `api` driver to `jwt`

```
'defaults' => [
    'guard' => 'api',
    'passwords' => 'users',
],
```

```
'guards' => [
    ...

    'api' => [
        'driver' => 'jwt',
        'provider' => 'users',
    ],
],
```

3. `/app/routes.php` - add a few basic authentication routes

```
Route::group([
    'prefix' => 'api'
], function () {

    $this->post('login', 'Auth\AuthController@login');
    $this->get('logout', 'Auth\AuthController@logout');
    
    Route::group([
        'prefix' => 'restricted',
        'middleware' => 'auth:api',
    ], function () {

        Route::get('/test', function () {
            return 'authenticated';
        });
    });
});
```

4. `/app/AuthController.php`

Appears to almost confirm out of the box to using the JWTGuard, while maintaining throttling on the API, except the second parameter of the call to `attempt`, is set to checking `$request->has('remember');` so in order to adhere to the JWT API.  I copied `AuthenticateUsers::login` into `AuthController` to edit the below lines.

```
public function login(Request $request)
{
    ...

    if ($token = Auth::guard($this->getGuard())->attempt($credentials)) {
        return $this->handleUserWasAuthenticated($request, $throttles, $token);
    }

    ...
}
```

5. `AuthenticatedUsers::handleUserWasAuthenticated` also needs to know about the `$token`

```
protected function handleUserWasAuthenticated(Request $request, $throttles, $token)
{
    if ($throttles) {
        $this->clearLoginAttempts($request);
    }

    if (method_exists($this, 'authenticated')) {
        return $this->authenticated($request, Auth::guard($this->getGuard())->user(), $token);
    }

    return redirect()->intended($this->redirectPath());
}
```

6. `AuthenticateUsers::handleUserWasAuthenticated` checks for  a method `authenticated`, which I added to `AuthController` to respond when authentication is successful

```
protected function authenticated($request, $user, $token)
{
    return response()->json([
        'request' => $request,
        'user'    => $user,
        'token' => $token,
    ]);
}
```

7. `AuthenticatesUsers::sendFailedLoginResponse` can be pulled up to `AuthController` to respond with JSON instead of redirecting failed authentication attempts

```
protected function sendFailedLoginResponse(Request $request)
{
    return response()->json([
        'request' => $request,
        'username' => $this->loginUsername(),
        'message'  => $this->getFailedLoginMessage(),
    ]);
}
```

8. `User` model needs to implement `Tymon\JWTAuth\Contracts\JWTSubject` (see [#260](https://github.com/tymondesigns/jwt-auth/issues/260#issuecomment-143683226), and [JWTSubject](https://github.com/tymondesigns/jwt-auth/blob/develop/src/Contracts/JWTSubject.php))

```
...

use Tymon\JWTAuth\Contracts\JWTSubject as AuthenticatableUserContract;

use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;

class User extends Model implements
    AuthenticatableContract,
    AuthorizableContract,
    CanResetPasswordContract,
    AuthenticatableUserContract
{
    use Authenticatable, Authorizable, CanResetPassword;

    ...

    /**
     * @return mixed
     */
    public function getJWTIdentifier()
    {
        return $this->getKey(); // Eloquent model method
    }

    /**
     * @return array
     */
    public function getJWTCustomClaims()
    {
        return [];
    }
}
```

9.  Test out whether authentication works by hitting the `/login` route with Postman

**Comment by MitchellMcKenna** [comment]
Thanks @mtpultz, would someone be able to post a similar guide to using JWTGuard with Lumen as well?

**Comment by mtpultz** [comment]
@tymondesigns seems worth adding a login action to the guard to allow for easy registration (this is more of a note).  If I have time on the weekend I'll have a look at it.

**Comment by ctadlock** [comment]
How is this coming along? I have it working but Im forced to use `dev-develop` so that I can use the JWTGuard login method.

**Comment by Luddinus** [comment]
+1

**Comment by michelecurletta** [comment]
+1, is that included into next first stable release (1.0)?

**Comment by 3amprogrammer** [comment]
Can someone explain me what is this parameter doing / what is going on in this line ?

```php
public function attempt(array $credentials = [], $login = true)
    {
        // ...
            return $login ? $this->login($user) : true;
        // ...
    }
```

**Comment by tymondesigns** [comment]
@3amprogrammer by default the method will return a token, but if you pass false as the second param then a boolean will be returned, indicating whether the credentials are valid

**Comment by 3amprogrammer** [comment]
@tymondesigns thanks for explanation. It is kinda strange cause when this method returns either a token or a boolean true we can be sure that the credentials where valid.

**Comment by HlaingTinHtun** [comment]
Can you show logout function in auth controller? @mtpultz

**Comment by ghprod** [comment]
@mtpultz thats really great .. also need logout example :+1:

**Comment by mtpultz** [comment]
@HlaingTinHtun and @ghprod,

Logout
=====

```
/**
 * Log the user out of the application.
 *
 * @return \Illuminate\Http\Response
 */
public function logout()
{
    Auth::guard($this->getGuard())->logout();

    // ...
}
```

This invokes JWTGuard's logout method, which invalidates the token, resets the user, and unsets the token.

JWTGuard's Logout
-----

```
/**
 * Logout the user, thus invalidating the token.
 *
 * @param  bool  $forceForever
 *
 * @return void
 */
public function logout($forceForever = false)
{
    $this->requireToken()->invalidate($forceForever);

    $this->user = null;
    $this->jwt->unsetToken();
}
```

**Comment by HlaingTinHtun** [comment]
@mtpultz Thanks. It works successfully

**Comment by mtpultz** [comment]
To keep the complete answer all in one place I've also added the logout to the [original post of the "guide"](https://github.com/tymondesigns/jwt-auth/issues/513#issuecomment-186087297).

**Comment by belohlavek** [comment]
@mtpultz I'm geting `Method [handle] does not exist.` with Lumen (using _dev-develop_) after following your instructions on how to configure _auth.php_.
This was the result of trying to apply the Middleware to my routes group (using Dingo/Api).
I guess that this happens because it's unable to find the middleware, but I really have no idea :sweat_smile: 

Any clue why this happens?

Thanks for reading :muscle:

**Comment by belohlavek** [comment]
Never mind, forgot to register the middleware, which is commented-out by default in Lumen. My bad :sweat:

**Comment by antonioreyna** [comment]
@mtpultz for that should i install the dev-master? using composer? thanks

**Comment by mtpultz** [comment]
Hi @antonioreyna, yah you'll need dev-master.  That's the branch the currently has the JWTGuard class.  If you look  in `/src` of the master branch you'll see that there is no JWTGuard.php file, but switching to dev-master you'll see it.

Cheers

**Comment by pouyaamreji** [comment]
I'm sorry for this noobish question but i don't get it after step 6. where can i find `AuthenticateUsers::login`?

**Comment by mtpultz** [comment]
Hi @pouyaamreji, 

The AuthenticatesUsers::login is a trait applied to the AuthController.  It is actually a composite trait since it is a trait within the AuthenticatesAndRegistersUsers trait that you see being used by the AuthController.  If you need to find these files, and your IDE doesn't provide a shortcut to indexed classes or methods, look at the use statement at the top of the file:

`use Illuminate\Foundation\Auth\AuthenticatesAndRegistersUsers;`

This is the mapping to the file in your composer vendors file so you'd look inside:

`vendors/larave/framework/src/Illuminate/Foundation/Auth`

In there you'll find the AuthenticatesAndRegistersUsers class and if you open it up you'll see the AuthenticatesUsers class being used, and you'll also see the AuthenticatesUsers file in the same folder.  Inside that is the login action you're trying to find.

**Comment by adstechnetwork** [comment]
Hi.  I am working on getting the guard functionality working.  If I install dev-master through Composer, the JWTGuard.php file is not included in source.  If I install dev-develop, however, JWTGuard is included, but  I get an error: Fatal error: Class 'Tymon\JWTAuth\Providers\JWTAuthServiceProvider' not found when I include it in app.php under service providers.  Any suggestions would be most appreciated.

**Comment by feeekkk** [comment]
@adstechnetwork if you take a look at the providers in the src directory, you'll see there is a `LumenServiceProvider` and a `LaravelServiceProvider` now.

**Comment by adstechnetwork** [comment]
Thank you @feeekkk . That worked.  However now, when I make a post to login I get the following reflection exception: Class Tymon\JWTAuth\Providers\JWT\NamshiAdapter does not exist.  

Any thoughts?

**Comment by feeekkk** [comment]
@adstechnetwork one of these should help you https://github.com/tymondesigns/jwt-auth/search?q=NamshiAdapter&type=Issues&utf8=%E2%9C%93

**Comment by adstechnetwork** [comment]
@feeekkk thank you very much.  That fixed that piece and I can keep  moving forward.  Most appreciated.

**Comment by adstechnetwork** [comment]
I have my application utilizing JWTGuard through the dev-devlop branch, and am trying to utilize the standard auth object such as Auth::user(), Auth::guest() etc. in my controllers, but they are returning incorrectly, as if they were looking for login credentials or session.  So if a user has passed a correct JWOT token, Auth::guest() still believes they are not logged in.  Is there an implementation of Auth::guest() that will work?  If not, what is the best way to approach this? Would there be better middleware declaration I can use in the routes.php file? Thank you very much for any kind assistance in advance.

**Comment by mariaczi** [comment]
Hi,

I don't know what I am doing wrong, but  I am handling register,login,logout in User Controller. I am using Laravel 5.3.

I have two issues:

**First issue**
I did copy AuthenticateUsers::login to UserController and changed it to
```
public function login(Request $request)
    {
        $this->validateLogin($request);

        // If the class is using the ThrottlesLogins trait, we can automatically throttle
        // the login attempts for this application. We'll key this by the username and
        // the IP address of the client making these requests into this application.
        if ($lockedOut = $this->hasTooManyLoginAttempts($request)) {
            $this->fireLockoutEvent($request);

            return $this->sendLockoutResponse($request);
        }

        $credentials = $this->credentials($request);

        if ($token = Auth::guard($this->getGuard())->attempt($credentials)) {
            return $this->handleUserWasAuthenticated($request, $throttles, $token);
        }

        // If the login attempt was unsuccessful we will increment the number of attempts
        // to login and redirect the user back to the login form. Of course, when this
        // user surpasses their maximum number of attempts they will get locked out.
        if (! $lockedOut) {
            $this->incrementLoginAttempts($request);
        }

        return $this->sendFailedLoginResponse($request);
    }
```

`$this->validateLogin`  and all others of course does not exist. What should I do?

**Second issue**
In User model I have following code:
```
namespace App;

use Illuminate\Notifications\Notifiable;
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Support\Facades\Validator;

use Tymon\JWTAuth\Contracts\JWTSubject as AuthenticatableUserContract;
use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;

class User extends Model implements
    AuthenticatableContract,
    AuthorizableContract,
    CanResetPasswordContract,
    AuthenticatableUserContract
{
    use Notifiable, Authenticatable, Authorizable, CanResetPassword;
```

Authenticable is giving errors already in IDE.

All implements: `Class must be declared abstract or implement methods sendPasswordResetNotification, getEmailForPasswordReset` 
and `Authenticatable` - `Trait expected, class found`

Can someone help?

**Comment by mtpultz** [comment]
Hi @mariaczi, so this post is really for Laravel 5.2 since 5.3 was only released 3 days ago, but that said it should probably be updated for 5.3 so if I have time I might drop something in this weekend since I have a new project that needs to be started up.  That said the gist is the same you have a token and you need to get it to the client so hopefully these points might help:

1. You're moving an authentication action out from the trait and placing it in the UserController, which seems to suggest you might have a few misunderstandings regarding OOP (though your edit seems to suggest you figured that out), but aside from that you need to move AuthenitcatesUsers::login up into the LoginController, which uses the AuthenticatesUsers trait so it overrides the traits method.  The UserController is not for authentication so for separation of concerns keep all authentication related actions within the controllers inside of  /http/Controllers/Auth.

1. Reviewing step 6 with regards to what is related to the JWTAuth package, which is collecting a token, and passing it along for the response you should look at JWTGuard::attempt() where you'll see it returns the token, which is collected and then should be passed on to sendLoginResponse since handleuserWasAuthetnicated doesn't exist anymore.  You'll probably want to drop $request->session()->regenerate() in sendLoginResponse (but I'm just guessing right now), and then jump to step 8 and implement authenticated and add the $token param.  This should give you a good start now that you can return the token.

1. For your second issue I'd suggest reading the docs I think you'll find the answers you're looking for on how to setup the User model outside of applying the JWTAuth package.

Hope this helps

**Comment by mariaczi** [comment]
@mtpulz I and probably others would be really grateful if you would find time for complete tutorial for 5.3 to let understand it correctly as 5.3 changed some main ideas behind.

**Comment by kylesean** [comment]
@mtpultz  Could you please show me your middleware file, you defined 'middleware' => 'auth:api', I don't know how to use the middleware . I write it like this:
```
<?php

namespace App\Http\Middleware;
use Closure;
use Tymon\JWTAuth\Facades\JWTAuth;
use Exception;
class authJWT
{
    public function handle($request, Closure $next)
    {
        try {
            $user = JWTAuth::toUser($request->input('token'));
        } catch (Exception $e) {
            if ($e instanceof \Tymon\JWTAuth\Exceptions\TokenInvalidException){
                return response()->json(['error'=>'Token is Invalid']);
            }else if ($e instanceof \Tymon\JWTAuth\Exceptions\TokenExpiredException){
                return response()->json(['error'=>'Token is Expired']);
            }else{
                return response()->json(['error'=>'Something is wrong']);
            }
        }
        return $next($request);
    }
}

```
But it didn't work. Hope you can help me .Thanks.Happy Mid-Autumn Festival

**Comment by mtpultz** [comment]
Hi @kylesean,  I actually don't use JWTAuth I use the JWTGuard in Laravel 5.2+ so I don't have to write any middleware like you are doing.

**Comment by mabasic** [comment]
Hi, I have written a lesson on getting this package working with Lumen 5.3 on my website [JSON Web Token Authentication for Lumen REBOOT](https://laravelista.com/lessons/json-web-token-authentication-for-lumen-reboot)

**Comment by mtpultz** [comment]
I've added an updated version for getting JWTGuard working with Laravel 5.3.x #860.

**Comment by morenDev** [comment]
I have such an error with laravel 5.2 
BadMethodCallException in JWTGuard.php line 405:
Method [handle] does not exist.

**Comment by mtpultz** [comment]
You might have to provide a bit more information like the version of JWT Auth that you're using, maybe the action that is calling JWTGuard, etc
